### ISSUE TITLE
Redesign movement AI as a CombatDirector-routed brain bank

### ISSUE THREAD

**Issue by Dudiebug** [Dudiebug/terminatorplus#9]
## Summary

Replace the current single/legacy movement brain path with a small bank of movement-only specialist brains selected by `CombatDirector` intent. The goal is to let custom loadouts dynamically switch movement policies as combat strategy changes, while keeping `CombatDirector` as the sole owner of weapons, item use, hotbar selection, blocking, crystals/anchors, potions, and other combat actions.

The core recommendation from the research is: train and route movement brains by **CombatDirector strategy branch family**, not by exact loadout and not by raw held item.

## Why this is needed

The combat side is already organized around `CombatDirector`: it owns per-tick combat decisions, weapon/behavior selection, distance checks, cooldowns, phases, and pending multi-tick plans. Movement is the weak part: the legacy path is mostly vector-to-target plus bunny-hop movement, with a small neural overlay that still leaks into combat authority through outputs such as `BLOCK`.

That creates two problems:

- movement policy is too broad and low-bandwidth for mechanically different combat styles such as mace, trident, elytra, wind charge, crystals, anchors, and close melee
- movement brains can accidentally compete with `CombatDirector` if they are allowed to issue combat actions

A routed movement-brain bank should improve positioning without creating a second combat owner.

## Proposed brain families

Start with a small, stable set of movement families rather than one brain per exact loadout:

- `general_fallback`
- `melee`
- `mace`
- `trident_ranged`
- `spear_melee`
- `mobility`
- `explosive_survival`

Do not mirror every `OpportunityScanner` tactical play as a separate brain. `scanner` is too broad as a routing label; scanner plays should be normalized into semantic movement families, with optional `playId` telemetry for later analysis.

## Runtime routing design

Introduce an explicit movement routing contract emitted by `CombatDirector`, for example:

```java
CombatIntent {
    MovementBranchFamily branchFamily;
    String playId; // optional/debug/telemetry
    DesiredRangeBand desiredRangeBand;
    boolean movementLocked;
    int lockTicksRemaining;
}
```

Routing precedence should be:

1. **Committed phase / movement lock**
   - Preserve the active brain through mace charging/airborne states, wind-charge plans, elytra glide/dive, combo commitments, and similar multi-tick actions.
   - Avoid mid-commit switching.

2. **Explicit `CombatIntent.branchFamily`**
   - `CombatDirector` decides the combat problem being solved.
   - The movement router selects the corresponding movement brain.

3. **Desired range / planned weapon role**
   - Use range bands as route modifiers and metrics, not as the primary owner of strategy.

4. **Held item as validator/tiebreaker only**
   - Held item is too weak as the primary signal because tridents, wind charges, elytra, offhand items, and pending plans can imply different movement roles.

## Movement-only contract

Movement brains may output locomotion only, such as:

- forward/back pressure
- strafe direction/intensity
- jump
- sprint
- crouch, if needed
- optional movement-state reporting such as retreating, circling, falling, holding, or closing

Movement brains must not directly call or decide:

- attack/punch
- item use
- shield/block
- hotbar selection
- projectile throws
- potion use
- crystal placement/detonation
- anchor placement/detonation
- cobweb/lava/firework/pearl/wind-charge actions
- `selectMaterial` or other inventory combat decisions

Those remain owned by `CombatDirector` and its behavior classes.

## Persistence design

Use a manifest plus per-brain files instead of one giant `ai/brain.json`.

Suggested layout:

```text
plugins/TerminatorPlus/ai/movement/manifest.json
plugins/TerminatorPlus/ai/movement/brains/general.json
plugins/TerminatorPlus/ai/movement/brains/melee.json
plugins/TerminatorPlus/ai/movement/brains/mace.json
plugins/TerminatorPlus/ai/movement/brains/trident_ranged.json
plugins/TerminatorPlus/ai/movement/brains/spear_melee.json
plugins/TerminatorPlus/ai/movement/brains/mobility.json
plugins/TerminatorPlus/ai/movement/brains/explosive_survival.json
```

Manifest should include:

- schema version
- observation schema hash
- action schema hash
- default brain name
- routing table version
- route-to-file mapping
- migration metadata

Each brain file should include:

- architecture metadata
- weights
- normalization statistics
- training build/version
- rollout/evaluation stats

Migration path: if only a legacy single brain exists, import it as `general_fallback` and alias all routes to it until specialized experts are trained.

## Training plan

Implement training in two stages:

1. **Branch-focused curriculum training**
   - close melee pressure
   - mace launch/airborne tracking/smash conversion
   - trident run-up/charge/release spacing
   - pearl/wind/elytra mobility
   - crystal/anchor safe-spacing and escape behavior

2. **Routed mixed-loadout self-play fine-tuning**
   - use the real `CombatDirector` router
   - include custom mixed loadouts
   - teach handoff timing and route transitions under pressure

Avoid fitness that only rewards survival time; that can teach stalling. Include route-specific success metrics.

## Evaluation plan

Compare an ablation ladder using the same `CombatDirector`, loadout pool, arenas, and seeds:

1. current hand-authored/legacy movement
2. one general movement brain
3. weapon-family routing
4. branch-family routing without commit latching
5. full branch-family routing with commit latching

Evaluate both default kits and custom mixed loadouts, including examples such as:

- sword/axe melee
- mace
- trident
- mobility-heavy
- crystal
- anchor
- `sword + mace + trident`
- `mace + wind charge + pearl`
- `trident + elytra + fireworks`
- `melee + crystal`

Primary metrics:

- win rate
- damage dealt minus damage taken
- self-damage / self-KO rate
- time spent inside active route desired range
- committed phase conversion rate
- mace-smash connect rate
- trident hit rate after charge
- mobility gap-close success rate
- safe explosive execution rate

Report confidence intervals or repeated-seed statistics rather than a single run.

## Implementation outline

1. Add movement intent types
   - Add `MovementBranchFamily` enum.
   - Add `CombatIntent` or equivalent routing DTO.
   - Normalize existing branch/debug labels into movement families.

2. Update `CombatDirector`
   - Emit explicit movement intent every tick.
   - Add movement lock/latch metadata for committed multi-tick phases.
   - Preserve `playId` for telemetry where scanner/opportunity plays are involved.

3. Add movement router
   - Select active brain by lock first, then branch family, then range/role fallback.
   - Default to `general_fallback` if an expert is missing or incompatible.
   - Add route-change telemetry.

4. Enforce the movement-only action contract
   - Remove combat-authority outputs such as legacy `BLOCK` from movement brains.
   - Ensure movement outputs cannot invoke inventory, item-use, attack, block, place, or detonate paths.

5. Add persistence layer
   - Add `ai/movement/manifest.json` support.
   - Add per-brain load/save support.
   - Add legacy single-brain import as `general_fallback`.
   - Validate observation/action schema hashes before loading specialized experts.

6. Update training pipeline
   - Allow training by movement family.
   - Add branch-focused scenario/curriculum setup.
   - Add routed mixed-loadout fine-tuning mode.
   - Store rollout stats per brain.

7. Add evaluation harness / commands
   - Run ablation variants with fixed seeds.
   - Emit route-specific metrics.
   - Include default and custom mixed loadout suites.

8. Preserve branch compatibility
   - Implement first on `mc-1.21.11`, then cherry-pick the concrete commit SHA to `mc-26.1.2` and `master`.
   - Build with `./gradlew build -q` on both active Minecraft branches.

## Acceptance criteria

- `CombatDirector` emits a semantic movement route/intent each tick.
- Movement brain selection is based on committed phase lock, then branch family, then fallback modifiers.
- Movement brains cannot execute combat actions directly.
- Existing behavior works with a single migrated `general_fallback` brain.
- Specialized brain files can be loaded independently from a manifest.
- Missing or incompatible specialized brains gracefully fall back to `general_fallback`.
- Custom mixed loadouts can switch movement families during a fight without loading a per-loadout brain.
- Evaluation can compare the ablation ladder with repeated seeds and route-specific metrics.
- `./gradlew build -q` succeeds on `mc-1.21.11` and `mc-26.1.2`.

## Notes

This should be treated as a larger architectural feature, likely split into implementation PRs:

1. routing contract and movement-only interface cleanup
2. router plus fallback behavior
3. persistence manifest and per-brain files
4. training/evaluation updates
5. specialized-brain training and tuning

**Comment by Dudiebug** [comment]
## Addendum: implementation gaps, reinforcement loadouts, and per-brain rewards

This issue should explicitly include the following requirements before implementation starts.

### Hard implementation gaps

1. **Add a side-effect-free planning pass before routing movement.**
   `CombatDirector.plan(...)` cannot only emit a loose string hint while `OpportunityScanner.scan(...)` and behavior classes execute the real action later. The movement router needs a typed, side-effect-free `CombatPlan`/`CombatIntent` first, then execution should consume that same plan. Otherwise movement can route as `neutral` while combat executes crystal, anchor, cobweb, pearl, mace, or scanner plays.

2. **Add a build/source completeness gate.**
   The movement-controller implementation must not rely on stale compiled classes under `build/classes`. Source files for movement/training APIs must exist and compile cleanly, including classes such as `MovementNetwork`, `MovementNetworkShape`, `MovementNetworkGenetics`, `MovementBrainPersistence`, `MovementTrainingConfig`, `MovementTrainingSnapshot`, and `CombatTrainingSnapshot`.

3. **Add real source `config.yml` defaults.**
   Since the plugin calls `saveDefaultConfig()`, movement/training defaults must live in source resources, not only generated/cached `build/resources/main/config.yml` output.

4. **Replace string/scalar branch coding.**
   Do not keep `String branchName` plus a single numeric `branchCode` as the final routing signal. Use a stable `MovementBranchFamily` enum for routing. If the selected family is exposed to the neural net, encode it as one-hot or explicit categorical fields so the model does not learn fake numeric ordering between unrelated families.

5. **Add a scanner-play-to-family mapping table.**
   `scanner` is not a movement family. Each `OpportunityScanner` play should map to a semantic movement family plus optional `playId`. Also add a `projectile_ranged` or `ranged_utility` family, because bow/crossbow/tipped-arrow/splash-projectile plays do not fit cleanly into `trident_ranged`.

6. **Define lock ownership and expiry.**
   `isCommitted` and `commitProgress` are not enough. Add fields such as `lockFamily`, `lockReason`, `lockUntilTick`, `interruptible`, and stale-lock clearing rules. This matters for mace charge/airborne, trident charge, wind-charge plans, combo chains, elytra dives, and consumable holds.

7. **Audit non-brain side effects outside `CombatDirector`.**
   Movement brains must not execute combat actions, but legacy paths still own some side effects such as shield blocking and environmental survival actions. Decide explicitly which remain legacy environmental safety and which move under `CombatDirector`/utility ownership.

8. **Expand the observation schema.**
   Kinematics plus desired range is not enough for mobility/explosive specialists. Add terrain/combat facts such as open sky, line of sight, nearby wall, target airborne/rising/falling, target blocking/eating/drinking, water/lava/cobweb/fire, blast danger, safe explosive distance, active route family, and `playId`.

9. **Make persistence atomic and failure-tolerant.**
   Manifest/per-brain writes should use temp files plus atomic replace where possible, backups, per-brain quarantine for bad files, version/schema mismatch handling, and “load valid subset plus fallback” semantics.

10. **Respect the Bukkit/NMS thread boundary during training.**
    Bukkit/NMS state should be captured on the main thread into immutable snapshots. Async training should consume only snapshots and pure network data.

### `/ai reinforcement` must automatically assign training loadouts

`/ai reinforcement` should not require a separate manual `/bot loadoutmix` step. When training movement-controller brains, the training session should automatically assign a weighted loadout to every spawned bot and record that loadout in telemetry/fitness snapshots.

Implementation requirements:

- Add a training config key such as `ai.training.loadout-mix: movement_balanced`.
- Add a weighted loadout sampler in the training pipeline, not only the `/bot loadoutmix` command path.
- Apply the sampled named loadout immediately after each training bot is created.
- Store the assigned loadout name/family in `CombatTrainingSnapshot` and movement rollout stats.
- Support per-family curriculum overrides, for example training only `mace`, then later running mixed self-play.
- Never let crystal/anchor/full-chaos kits dominate the population; they should be represented but capped.

### Proposed default training loadout mix

Use existing named loadouts where possible. Suggested weighted `movement_balanced` mix, total weight 100:

```yaml
movement_balanced:
  sword: 12
  axe: 12
  smp: 12
  pot: 8
  mace: 10
  spear: 8
  trident: 8
  windcharge: 6
  skydiver: 5
  hybrid: 6
  vanilla: 5
  pvp: 3
  crystalpvp: 3
  anchorbomb: 2
```

This keeps `crystalpvp + anchorbomb + pvp` at 8% total. That is enough to teach explosive/survival movement without turning every generation into crystal chaos.

Suggested curriculum mixes:

```yaml
melee_curriculum:
  sword: 20
  axe: 20
  smp: 25
  pot: 15
  spear: 10
  hybrid: 10

mace_curriculum:
  mace: 45
  hybrid: 20
  windcharge: 15
  vanilla: 10
  pvp: 5
  skydiver: 5

trident_curriculum:
  trident: 40
  spear: 25
  skydiver: 15
  hybrid: 10
  vanilla: 10

mobility_curriculum:
  windcharge: 25
  skydiver: 25
  hybrid: 20
  trident: 10
  mace: 10
  pvp: 5
  vanilla: 5

explosive_survival_curriculum:
  crystalpvp: 20
  anchorbomb: 15
  vanilla: 20
  pvp: 15
  hybrid: 10
  windcharge: 10
  sword: 10
```

For normal `/ai reinforcement` movement training, default to `movement_balanced`. Use the curriculum mixes only when explicitly training a branch family.

### Loadouts to preserve or add

Preserve these existing named loadouts as the first training pool:

- `sword`: simple sword/shield melee baseline
- `axe`: shield-break axe pressure
- `smp`: sword plus axe, no explosive/mace chaos
- `pot`: sword plus splash-healing sustain
- `mace`: mace plus wind-charge launch pairing
- `spear`: trident used as close-range spear/melee fallback
- `trident`: ranged trident behavior
- `windcharge`: wind-charge movement/zoning
- `skydiver`: elytra/firework/trident mobility
- `hybrid`: sword plus mace plus trident plus wind charge
- `vanilla`: broad non-elytra full-kit baseline
- `pvp`: everything-bagel stress kit, low weight only
- `crystalpvp`: crystal specialist, low weight only
- `anchorbomb`: anchor specialist, low weight only

If new training-only loadouts are added later, keep them small and branch-specific rather than adding many near-duplicates. Useful additions would be:

- `melee_basic`: sword/axe/shield with no pearls, crystals, mace, trident, or wind charges
- `mace_wind`: mace plus offhand wind charges, minimal secondary weapon
- `trident_ranged`: trident plus modest armor, no sword dominance
- `spear_only`: trident-only close-range kit
- `mobility_light`: pearls/wind/elytra/fireworks with weak melee backup
- `explosive_light`: small crystal/anchor kit with strict resource limits

### Per-brain reward profiles

Each brain family should have a different reward profile. A single global fitness function will blur the specialties.

- `general_fallback`: balanced damage delta, survival, low no-op/fallback rate, sane range control, low oscillation, no route thrash.
- `melee`: reward time in melee threat range, legal sprint-hit/crit setup, hit conversion, sweep setup when sword is selected, and damage dealt. Penalize jump spam, backing out of range, and whiffing into iframes.
- `mace`: reward successful charge/launch/airborne/smash phase conversion, target tracking while airborne, smash connect rate, and preserving mace selection through commitment. Penalize failed launch, landing without a smash attempt, self-fall damage, and mid-commit route switches.
- `trident_ranged`: reward maintaining throw range/line of sight, charge completion, hit rate after release, and spacing that avoids collapse into melee. Penalize charging out of range, blocked line throws, and over-retreating.
- `spear_melee`: reward close trident pressure, stable melee spacing, hit conversion, and fallback behavior when no sword/axe exists. Penalize trying to behave like ranged trident at point blank.
- `mobility`: reward gap close, escape from bad range, vertical setup when useful, route handoff success, and maintaining recoverable velocity. Penalize stalling, flying away forever, unsafe landings, and losing line of sight unnecessarily.
- `explosive_survival`: reward target damage from crystals/anchors, safe blast spacing, post-setup escape, and low self-damage/self-KO. Penalize detonating inside unsafe radius or failing to retreat after placement.
- `projectile_ranged` / `ranged_utility`: reward line-of-sight control, lateral strafing, projectile hit/interrupt success, and useful distance bands. Penalize wasting arrows/potions/fireworks with no tactical gain.

The evaluation harness should report per-family fitness components, not just one aggregate score, so a model cannot hide bad mace or explosive behavior behind good generic survival.
